src/main/java/bio/terra/stairway/Database.java

```diff
@@ -264,6 +239,107 @@ public void complete(FlightContext flightContext) {
         return flightList;
     }
 
+    /**
+     * Return flight state for a single flight
+     *
+     * @param flightId
+     * @return FlightState for the flight
+     */
+    public FlightState getFlightState(String flightId) {
+        final String sqlOneFlight = "SELECT flightid, submit_time, input_parameters," +
+                " completed_time, output_parameters, status, error_message" +
+                " FROM " + FLIGHT_TABLE +
+                " WHERE flightid = :flightid";
+
+        try (Connection connection = dataSource.getConnection();
+             NamedParameterPreparedStatement oneFlightStatement =
+                     new NamedParameterPreparedStatement(connection, sqlOneFlight)) {
+
+            oneFlightStatement.setString("flightid", flightId);
+
+            try (ResultSet rs = oneFlightStatement.getPreparedStatement().executeQuery()) {
+                List<FlightState> flightStateList = makeFlightStateList(rs);
+                if (flightStateList.size() == 0) {
+                    throw new FlightNotFoundException("Flight not found: " + flightId);
+                }
+                if (flightStateList.size() > 1) {
+                    throw new DatabaseOperationException("Multiple flights with the same id?!");
+                }
+                return flightStateList.get(0);
+            }
+
+        } catch (SQLException ex) {
+            throw new DatabaseOperationException("Failed to get flight", ex);
+        }
+    }
+
+    public List<FlightState> getFlights(int offset, int limit) {
+        final String sqlFlightRange = "SELECT flightid, submit_time, input_parameters," +
```

Comment:
when someone is getting a list of flight states, would they want to know if the flight is in the doing or undoing state if it is still running?

Reply:
Good question. I considered piling a lot of state into the flight state return, but thought that the main use would be our debugging and not the UI. For our debugging, we can look at the (to be implemented) log or directly at the database.

It would certainly be straightforward to add if we decide we want to return more state about flight progress.